cpp/test/sparse/sddmm.cu

```diff
@@ -0,0 +1,425 @@
+      c_indptr_d.data(),
+      c_indices_d.data(),
+      params.m,
+      params.n,
+      static_cast<IndexType>(c_indices_d.size()));
+
+    auto c = raft::make_device_csr_matrix_view<ValueType>(c_data_d.data(), c_structure);
+
+    RAFT_CUDA_TRY(cudaStreamSynchronize(resource::get_cuda_stream(handle)));
+
+    auto op_a = raft::linalg::Operation::NON_TRANSPOSE;
+    auto op_b = !(std::is_same_v<LayoutPolicyA, LayoutPolicyB>)
+                  ? raft::linalg::Operation::NON_TRANSPOSE
+                  : raft::linalg::Operation::TRANSPOSE;
+
+    raft::sparse::linalg::sddmm(handle,
+                                a,
+                                b,
+                                c,
+                                op_a,
+                                op_b,
+                                raft::make_host_scalar_view<ValueType>(&params.alpha),
+                                raft::make_host_scalar_view<ValueType>(&params.beta));
+
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+    RAFT_CUDA_TRY(cudaDeviceSynchronize());
+
+    ASSERT_TRUE(raft::devArrMatch<ValueType>(c_expected_data_d.data(),
+                                             c.get_elements().data(),
+                                             params.c_indices.size(),
+                                             raft::CompareApprox<ValueType>(1e-6f),
+                                             stream));
+  }
+
+  raft::resources handle;
+  SDDMMInputs<ValueType, IndexType> params;
+  cudaStream_t stream;
+
+  rmm::device_uvector<ValueType> a_data_d;
+  rmm::device_uvector<ValueType> b_data_d;
+
+  rmm::device_uvector<IndexType> c_indptr_d;
+  rmm::device_uvector<IndexType> c_indices_d;
+  rmm::device_uvector<ValueType> c_data_d;
+
+  rmm::device_uvector<ValueType> c_expected_data_d;
+};
+
+using SDDMMTestF_Row_Col = SDDMMTest<float, int, raft::row_major, raft::col_major>;
+TEST_P(SDDMMTestF_Row_Col, Result) { Run(); }
+
+using SDDMMTestF_Col_Row = SDDMMTest<float, int, raft::col_major, raft::row_major>;
+TEST_P(SDDMMTestF_Col_Row, Result) { Run(); }
+
+using SDDMMTestF_Row_Row = SDDMMTest<float, int, raft::row_major, raft::row_major>;
+TEST_P(SDDMMTestF_Row_Row, Result) { Run(); }
+
+using SDDMMTestF_Col_Col = SDDMMTest<float, int, raft::col_major, raft::col_major>;
+TEST_P(SDDMMTestF_Col_Col, Result) { Run(); }
+
+using SDDMMTestD_Row_Col = SDDMMTest<double, int, raft::row_major, raft::col_major>;
+TEST_P(SDDMMTestD_Row_Col, Result) { Run(); }
+
+using SDDMMTestD_Col_Row = SDDMMTest<double, int, raft::col_major, raft::row_major>;
+TEST_P(SDDMMTestD_Col_Row, Result) { Run(); }
+
+using SDDMMTestD_Row_Row = SDDMMTest<double, int, raft::row_major, raft::row_major>;
+TEST_P(SDDMMTestD_Row_Row, Result) { Run(); }
+
+using SDDMMTestD_Col_Col = SDDMMTest<double, int, raft::col_major, raft::col_major>;
+TEST_P(SDDMMTestD_Col_Col, Result) { Run(); }
+
+const std::vector<SDDMMInputs<float, int>> sddmm_inputs_row_col_f = {
+  {
+    4,
+    4,
+    3,
+    1.0,
```

Comment:
There are a few places where we do this because it's really hard to generate the outputs for automating the comparisons. The problem with hardcoding these things is that it's non-trivial to add new test cases and so when a user comes comes to us with an issue where they are seeing strange results with some sparsities or specific inputs, we couldn't otherwise be able to quickly reproduce in a test once we fix the issue. 

I think it's easy enough to automate the`expected` outputs, though- since you are already creating a mask, just copy out the code from your micro benchmarks that creates a random mask, generate two input dense arrays (using raft::random::make_blobs), compute the pairwise distances between A and B, and then copy in the rows/cols of the results to your "expected output" parse structure. 

Given how much work went into benchmarking, I'd also highly suggest we commit your benchmarking code with these changes also. Since it was already written, I can't express enough how convenient it is to be able to load up a simple microbenchmark when someone finds a specific case to test. I understand this version is a lightweight wrapper around the SDDMM from cusparse, but that's likely not always going to be the case and so automated tests and benchmarks helps us evolve the code over time (and back-ends).

Reply:
Make sense!